epidose/device/wps_scanner_d.sh

```diff
@@ -0,0 +1,134 @@
+UTIL="$(dirname "$0")/util.sh"
+
+# Source common functionality (logging, WiFi)
+# shellcheck source=epidose/device/util.sh
+SERVER_URL="http://ha-server:5010" . "$UTIL"
+
+# Light up LED lights to indicate network status messages.
+# precondition: none
+# Returns an exit code of 0 if a network connection is established,
+# while exit code 1 denotes a failed attempt to connect to a WPS network,
+# exit code 2 shows that there are no available WPS networks in the area,
+# and value 3 refers to an invalid LED color option.
+wps_led_blink()
+{
+  run_python device_io --"$1"-on
+  case "$1" in
+	"green") log "Obtained network connection"; return 0;;
+	"red") log "Failed to connect to a network"; return 1;;
+	"orange") log "No WPS network found in the area"; return 2;;
+	*) log "The given LED option is not available"return 3;;
+  esac
+  sleep 3
+  run_python device_io --"$1"-off
+}
+
+# Connect to a known network if available,
+# otherwise try to connect to the network with
+# the stongest signal by using WPS.
+# precondition: WiFi should be turned on.
+# Returns exit codes based on the return codes
+# of the wps_led_blink function
+wps_connect()
+{
+  network_status=$(ifconfig wlan0 | grep "inet ")
+
+  # Check if device is connected to any network
+  if [ ! -z "$network_status" ] ; then
+    # If connected to a network then return
+    return 0
+  fi
+
+  log "Initiating WPS connection"
+ 
+  # Suspend watchdog's LED activities
+  touch "$SUSPEND_WATCHDOG_LED"
+  
+  # Get WPS network with the strongest signal
+  get_network_with_strongest_signal=$(wpa_cli -i wlan0 scan_results | grep WPS | sort -r -k3 | tail -1)
+  if [ -z "$get_network_with_strongest_signal" ] ; then
+    # If there is no available WPS network in the area, turn on the orange LED light	    
+    wps_led_blink orange
+    exit_code=$?
+
+    # Release watchdog
+    rm "$SUSPEND_WATCHDOG_LED"
+    return "$exit_code"
+  fi	  
+
+  # Get bssid of the strongest network signal and connect to it 
+  echo "$get_network_with_strongest_signal" | awk '{print $1}' | xargs wpa_cli -i wlan0 wps_pbc
+     
+  # Get network name
+  network_name=$(echo "$get_network_with_strongest_signal" | awk '{print $NF}')
+  log "Trying to connect to $network_name"
+    
+  # Obtain dynamic IP address
+  dhclient wlan0
+
+  get_ip_address=$(ifconfig wlan0 | grep "inet ")
+    
+  # Check whether the connection is established and light up the LED accordingly
+  if [ -z "$get_ip_address" ] ; then
+    # If failed to connect to a network, turn on the red LED light
+    wps_led_blink red
+    exit_code=$?
+  else
+    # If connected to a network, turn on the green LED light
+    wps_led_blink green
+    exit_code=$?
+  fi    
+  
+  # Release watchdog
+  rm "$SUSPEND_WATCHDOG_LED"
+    
+  return "$exit_code"
+}
+
+while : ; do
+  # Wait for the Wifi button to be pressed
+  run_python device_io --wifi-wait
+  wifi_acquire
+  if wps_connect; then
+    log "Updating Cuckoo filter" 
+    if get_new_filter; then
```

Comment:
There's no need (and it is risky due to race conditions) to create a second instance of this function. Simply killing the *sleep* will cause it to be called, no?

Reply:
Yes, I have removed the get_new_filter function and now it only kills the sleep process.
However, for some reason the update_filter daemon, after the sleep process that is killed,
it exits without any error message.
Any idea of what the issue might be?

Reply:
To debug this issue I constructed a small test script
```
#!/bin/sh
while : ; do
  sleep 5 &
  pid=$!
  echo $pid
  wait $pid
done
```

This works fine, until you add `set -e` at its beginning.  Checking the `wait` exit code e.g. with `  wait $pid || echo "Killed"` addresses the problem.  You should add an appropriate log statement instead of echo.

Reply:
I had no clue that set -e was causing this!
I have applied all the requested changes and I have submitted them.
Please let me know if more changes are required.